graphix/open_graph.py

```diff
@@ -0,0 +1,248 @@
+        >>> measurements = {i: Measurement(0.5 * i, Plane.XY) for i in range(2)}
+        >>> inputs = [0]
+        >>> outputs = [2]
+        >>>
+        >>> og = OpenGraph(g, measurements, inputs, outputs)
+        >>> og.inputs == inputs
+        True
+        """
+        inputs = [i for i in self._inside.nodes(data=True) if "is_input" in i[1]]
+        sorted_inputs = sorted(inputs, key=lambda x: x[1]["input_order"])
+
+        # Returns only the input ids
+        return [i[0] for i in sorted_inputs]
+
+    @property
+    def outputs(self) -> list[int]:
+        """Returns the outputs of the graph sorted by their position.
+
+        Example
+        ------
+        >>> import networkx as nx
+        >>> from graphix.open_graph import OpenGraph, Measurement
+        >>>
+        >>> g = nx.Graph([(0, 1), (1, 2), (2, 0)])
+        >>> measurements = {i: Measurement(0.5 * i, Plane.XY) for i in range(2)}
+        >>> inputs = [0]
+        >>> outputs = [2]
+        >>>
+        >>> og = OpenGraph(g, measurements, inputs, outputs)
+        >>> og.outputs == outputs
+        True
+        """
+        outputs = [i for i in self._inside.nodes(data=True) if "is_output" in i[1]]
+        sorted_outputs = sorted(outputs, key=lambda x: x[1]["output_order"])
+        output_node_ids = [i[0] for i in sorted_outputs]
+        return output_node_ids
+
+    @property
+    def measurements(self) -> dict[int, Measurement]:
+        """Returns a dictionary which maps each node to its measurement. Output
+        nodes are not measured and hence are not included in the dictionary.
+
+        Example
+        ------
+        >>> import networkx as nx
+        >>> from graphix.open_graph import OpenGraph, Measurement
+        >>>
+        >>> g = nx.Graph([(0, 1), (1, 2), (2, 0)])
+        >>> measurements = {i: Measurement(0.5 * i, Plane.XY) for i in range(2)}
+        >>> inputs = [0]
+        >>> outputs = [2]
+        >>>
+        >>> og = OpenGraph(g, measurements, inputs, outputs)
+        >>> og.measurements == {
+        ...     0: Measurement(0.0, Plane.XY),
+        ...     1: Measurement(0.5, Plane.XY),
+        ... }
+        True
+        """
+        return {n[0]: n[1]["measurement"] for n in self._inside.nodes(data=True) if "measurement" in n[1]}
+
+    @classmethod
+    def from_pattern(cls, pattern: Pattern) -> OpenGraph:
+        """Initialises an OpenGraph based on a measurement pattern"""
+        g = nx.Graph()
+        nodes, edges = pattern.get_graph()
+        g.add_nodes_from(nodes)
+        g.add_edges_from(edges)
+
+        inputs = pattern.input_nodes
+        outputs = pattern.output_nodes
+
+        meas_planes = pattern.get_meas_plane()
+        meas_angles = pattern.get_angles()
+        meas = {node: Measurement(meas_angles[node], meas_planes[node]) for node in meas_angles}
+
+        return cls(g, meas, inputs, outputs)
+
+    def to_pattern(self) -> Pattern:
+        """Converts the Open graph into a pattern."""
```

Comment:
please specifically state that the conversion only succeeds if either flow, gflow or pauli flow exist, and that the generated pattern is induced by maximally-delayed flow.